README.md
# GitStatusBg (Git Status Background)

Visual Studio Code extension that sets the background color of the editor to indicate the Git status of the file.

## Features

Color the background of the editor based on the Git status of that file.
This is to help you easily identify these files, and help avoid making changes in the wrong place or when they may be overwritten.

The following statuses are supported

* Untracked
* Modified (locally)
* Behind the remote version

The images below show open modified and untracked files in the editor.

![Example using dark theme](assets/screenshot-dark.png)
![Example using light theme](assets/screenshot-light.png)

If you would prefer to use different colors (or they don't work with your chosen theme), the colors can be changed in settings to whatever you wish.

## Requirements

The default (built-in) GIT provider is used to get the status of files. If you're not using this, or if you're not using GIT as a source control system, this extension isn't going to be of much use to you.

## Extension Settings

Settings exist to adjust the tint color that is applied to the background. Adjust based on your preference.

* `gitstatusbg.untrackedFileBackground`: the tint color to use for untracked files
* `gitstatusbg.modifiedFileBackground`: the tint color to use for modified files
* `gitstatusbg.behindRemoteFileBackground`: the tint color to use for files that have been modified remotely but those changes have not been pulled locally

## Known Issues

None yet. If you find something wrong or have a suggestion, please [raise an issue](https://github.com/mrlacey/GitStatusBg/issues).

## Release Notes

### 1.1.0

* Added coloring of files that are behind the remote.
* Fixed bug where coloring may not be updated after saving a document.

### 1.0.0

Initial release.


CHANGELOG.md
# Change Log

All notable changes to the "gitstatusbg" extension will be documented in this file.

## 1.1.4

- Updated dependencies for security vulnerability fixes.

## 1.1.3

- Updated dependencies for security vulnerability fixes.

## 1.1.2

- Updated dependencies for security vulnerability fixes.

## 1.1.1

- Updated dependencies for security vulnerability fixes.

## 1.1.0

- Initial release


src/api/git.d.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { Uri, Event, Disposable, ProviderResult } from 'vscode';
export { ProviderResult } from 'vscode';

export interface Git {
	readonly path: string;
}

export interface InputBox {
	value: string;
}

export const enum RefType {
	Head,
	RemoteHead,
	Tag
}

export interface Ref {
	readonly type: RefType;
	readonly name?: string;
	readonly commit?: string;
	readonly remote?: string;
}

export interface UpstreamRef {
	readonly remote: string;
	readonly name: string;
}

export interface Branch extends Ref {
	readonly upstream?: UpstreamRef;
	readonly ahead?: number;
	readonly behind?: number;
}

export interface Commit {
	readonly hash: string;
	readonly message: string;
	readonly parents: string[];
	readonly authorDate?: Date;
	readonly authorName?: string;
	readonly authorEmail?: string;
	readonly commitDate?: Date;
}

export interface Submodule {
	readonly name: string;
	readonly path: string;
	readonly url: string;
}

export interface Remote {
	readonly name: string;
	readonly fetchUrl?: string;
	readonly pushUrl?: string;
	readonly isReadOnly: boolean;
}

export const enum Status {
	INDEX_MODIFIED,
	INDEX_ADDED,
	INDEX_DELETED,
	INDEX_RENAMED,
	INDEX_COPIED,

	MODIFIED,
	DELETED,
	UNTRACKED,
	IGNORED,
	INTENT_TO_ADD,

	ADDED_BY_US,
	ADDED_BY_THEM,
	DELETED_BY_US,
	DELETED_BY_THEM,
	BOTH_ADDED,
	BOTH_DELETED,
	BOTH_MODIFIED
}

export interface Change {

	/**
	 * Returns either `originalUri` or `renameUri`, depending
	 * on whether this change is a rename change. When
	 * in doubt always use `uri` over the other two alternatives.
	 */
	readonly uri: Uri;
	readonly originalUri: Uri;
	readonly renameUri: Uri | undefined;
	readonly status: Status;
}

export interface RepositoryState {
	readonly HEAD: Branch | undefined;
	readonly refs: Ref[];
	readonly remotes: Remote[];
	readonly submodules: Submodule[];
	readonly rebaseCommit: Commit | undefined;

	readonly mergeChanges: Change[];
	readonly indexChanges: Change[];
	readonly workingTreeChanges: Change[];

	readonly onDidChange: Event<void>;
}

export interface RepositoryUIState {
	readonly selected: boolean;
	readonly onDidChange: Event<void>;
}

/**
 * Log options.
 */
export interface LogOptions {
	/** Max number of log entries to retrieve. If not specified, the default is 32. */
	readonly maxEntries?: number;
	readonly path?: string;
}

export interface CommitOptions {
	all?: boolean | 'tracked';
	amend?: boolean;
	signoff?: boolean;
	signCommit?: boolean;
	empty?: boolean;
}

export interface BranchQuery {
	readonly remote?: boolean;
	readonly contains?: string;
}

export interface Repository {

	readonly rootUri: Uri;
	readonly inputBox: InputBox;
	readonly state: RepositoryState;
	readonly ui: RepositoryUIState;

	getConfigs(): Promise<{ key: string; value: string; }[]>;
	getConfig(key: string): Promise<string>;
	setConfig(key: string, value: string): Promise<string>;
	getGlobalConfig(key: string): Promise<string>;

	getObjectDetails(treeish: string, path: string): Promise<{ mode: string, object: string, size: number }>;
	detectObjectType(object: string): Promise<{ mimetype: string, encoding?: string }>;
	buffer(ref: string, path: string): Promise<Buffer>;
	show(ref: string, path: string): Promise<string>;
	getCommit(ref: string): Promise<Commit>;

	clean(paths: string[]): Promise<void>;

	apply(patch: string, reverse?: boolean): Promise<void>;
	diff(cached?: boolean): Promise<string>;
	diffWithHEAD(): Promise<Change[]>;
	diffWithHEAD(path: string): Promise<string>;
	diffWith(ref: string): Promise<Change[]>;
	diffWith(ref: string, path: string): Promise<string>;
	diffIndexWithHEAD(): Promise<Change[]>;
	diffIndexWithHEAD(path: string): Promise<string>;
	diffIndexWith(ref: string): Promise<Change[]>;
	diffIndexWith(ref: string, path: string): Promise<string>;
	diffBlobs(object1: string, object2: string): Promise<string>;
	diffBetween(ref1: string, ref2: string): Promise<Change[]>;
	diffBetween(ref1: string, ref2: string, path: string): Promise<string>;

	hashObject(data: string): Promise<string>;

	createBranch(name: string, checkout: boolean, ref?: string): Promise<void>;
	deleteBranch(name: string, force?: boolean): Promise<void>;
	getBranch(name: string): Promise<Branch>;
	getBranches(query: BranchQuery): Promise<Ref[]>;
	setBranchUpstream(name: string, upstream: string): Promise<void>;

	getMergeBase(ref1: string, ref2: string): Promise<string>;

	status(): Promise<void>;
	checkout(treeish: string): Promise<void>;

	addRemote(name: string, url: string): Promise<void>;
	removeRemote(name: string): Promise<void>;
	renameRemote(name: string, newName: string): Promise<void>;

	fetch(remote?: string, ref?: string, depth?: number): Promise<void>;
	pull(unshallow?: boolean): Promise<void>;
	push(remoteName?: string, branchName?: string, setUpstream?: boolean): Promise<void>;

	blame(path: string): Promise<string>;
	log(options?: LogOptions): Promise<Commit[]>;

	commit(message: string, opts?: CommitOptions): Promise<void>;
}

export interface RemoteSource {
	readonly name: string;
	readonly description?: string;
	readonly url: string | string[];
}

export interface RemoteSourceProvider {
	readonly name: string;
	readonly icon?: string; // codicon name
	readonly supportsQuery?: boolean;
	getRemoteSources(query?: string): ProviderResult<RemoteSource[]>;
	publishRepository?(repository: Repository): Promise<void>;
}

export interface Credentials {
	readonly username: string;
	readonly password: string;
}

export interface CredentialsProvider {
	getCredentials(host: Uri): ProviderResult<Credentials>;
}

export type APIState = 'uninitialized' | 'initialized';

export interface API {
	readonly state: APIState;
	readonly onDidChangeState: Event<APIState>;
	readonly git: Git;
	readonly repositories: Repository[];
	readonly onDidOpenRepository: Event<Repository>;
	readonly onDidCloseRepository: Event<Repository>;

	toGitUri(uri: Uri, ref: string): Uri;
	getRepository(uri: Uri): Repository | null;
	init(root: Uri): Promise<Repository | null>;

	registerRemoteSourceProvider(provider: RemoteSourceProvider): Disposable;
	registerCredentialsProvider(provider: CredentialsProvider): Disposable;
}

export interface GitExtension {

	readonly enabled: boolean;
	readonly onDidChangeEnablement: Event<boolean>;

	/**
	 * Returns a specific API version.
	 *
	 * Throws error if git extension is disabled. You can listed to the
	 * [GitExtension.onDidChangeEnablement](#GitExtension.onDidChangeEnablement) event
	 * to know when the extension becomes enabled/disabled.
	 *
	 * @param version Version number.
	 * @returns API instance
	 */
	getAPI(version: 1): API;
}

export const enum GitErrorCodes {
	BadConfigFile = 'BadConfigFile',
	AuthenticationFailed = 'AuthenticationFailed',
	NoUserNameConfigured = 'NoUserNameConfigured',
	NoUserEmailConfigured = 'NoUserEmailConfigured',
	NoRemoteRepositorySpecified = 'NoRemoteRepositorySpecified',
	NotAGitRepository = 'NotAGitRepository',
	NotAtRepositoryRoot = 'NotAtRepositoryRoot',
	Conflict = 'Conflict',
	StashConflict = 'StashConflict',
	UnmergedChanges = 'UnmergedChanges',
	PushRejected = 'PushRejected',
	RemoteConnectionError = 'RemoteConnectionError',
	DirtyWorkTree = 'DirtyWorkTree',
	CantOpenResource = 'CantOpenResource',
	GitNotFound = 'GitNotFound',
	CantCreatePipe = 'CantCreatePipe',
	CantAccessRemote = 'CantAccessRemote',
	RepositoryNotFound = 'RepositoryNotFound',
	RepositoryIsLocked = 'RepositoryIsLocked',
	BranchNotFullyMerged = 'BranchNotFullyMerged',
	NoRemoteReference = 'NoRemoteReference',
	InvalidBranchName = 'InvalidBranchName',
	BranchAlreadyExists = 'BranchAlreadyExists',
	NoLocalChanges = 'NoLocalChanges',
	NoStashFound = 'NoStashFound',
	LocalChangesOverwritten = 'LocalChangesOverwritten',
	NoUpstreamBranch = 'NoUpstreamBranch',
	IsInSubmodule = 'IsInSubmodule',
	WrongCase = 'WrongCase',
	CantLockRef = 'CantLockRef',
	CantRebaseMultipleBranches = 'CantRebaseMultipleBranches',
	PatchDoesNotApply = 'PatchDoesNotApply',
	NoPathFound = 'NoPathFound',
	UnknownPath = 'UnknownPath',
}

src/extension.ts
import * as vscode from "vscode";

import { GitExtension, API, Status } from "./api/git";

// alpha of 0.05(5%) should allow for subtle overlay on dark or light backgrounds
// These are also defined in the configuration
let COLOR_MODIFIED: string | undefined = "rgba(255, 165, 0, 0.05)";
let COLOR_UNTRACKED: string | undefined = "rgba(71, 255, 25, 0.05)";
let COLOR_BEHIND: string | undefined = "rgba(255, 0, 0, 0.05)";

const enum XStatus {
  CHANGED_ON_SERVER = 100,
}

let git: API | undefined = undefined;

// Keep track of current decorations so that they can be removed
let currentDecorations: {
  [path: string]: vscode.TextEditorDecorationType | undefined;
} = {};

let fileStatusCache: {
  [path: string]: Status | XStatus | undefined;
} = {};

export async function activate(context: vscode.ExtensionContext) {
  const ext = vscode.extensions.getExtension<GitExtension>("vscode.git");

  await ext?.activate();
  git = ext!.exports.getAPI(1);

  git.onDidChangeState;

  context.subscriptions.push(
    // Handle switching (or opening) document tabs
    vscode.window.onDidChangeActiveTextEditor(onDidChangeActiveTextEditor),

    // Handle scrolling in the document.
    vscode.window.onDidChangeTextEditorVisibleRanges(
      onDidChangeTextEditorVisibleRanges
    ),

    // File saved so pick up any changes because of saving
    vscode.workspace.onDidSaveTextDocument(onFileSaved),

    // Potentially large number of files changed - just reset the status cache (lazy but effective)
    vscode.workspace.onDidChangeWorkspaceFolders((e) => {
      fileStatusCache = {};
    }),
    vscode.workspace.onDidCreateFiles((e) => {
      fileStatusCache = {};
    }),
    vscode.workspace.onDidDeleteFiles((e) => {
      fileStatusCache = {};
    }),
    vscode.workspace.onDidRenameFiles((e) => {
      fileStatusCache = {};
    })
  );

  // Check for initial document
  RefreshCurrentEditor();
}

function RefreshCurrentEditor() {
  onDidChangeActiveTextEditor(vscode.window.activeTextEditor);
}

function onFileSaved(e: vscode.TextDocument) {
  // Document may have been colored before saving but shouldn't be colored after
  // - therefore remove any existing coloring
  var existing = currentDecorations[e.fileName];

  if (existing) {
    vscode.window?.activeTextEditor?.setDecorations(existing, []);
  }

  // Delete the cached status because it may have changed because saving
  fileStatusCache[e.fileName] = undefined;

  // Update the coloring of the document
  SetColorToStatus(vscode.window.activeTextEditor);
}

function onDidChangeTextEditorVisibleRanges(
  e: vscode.TextEditorVisibleRangesChangeEvent
) {
  let changedRanges: vscode.Range[] = [];

  changedRanges = changedRanges.concat(e.visibleRanges);

  SetColorToStatus(e.textEditor, changedRanges);
}

function onDidChangeActiveTextEditor(editor?: vscode.TextEditor) {
  if (editor === undefined) {
    return;
  }

  ReloadConfigurationInCaseChanged();

  // Clear the cached state of the file just switched to
  // - this allows for state changes of open files while another tab was displayed
  fileStatusCache[editor.document.fileName] = undefined;

  SetColorToStatus(editor);
}

function ReloadConfigurationInCaseChanged() {
  var config = vscode.workspace.getConfiguration();

  let untracked: any = config.get("gitstatusbg.untrackedFileBackground");

  if (untracked) {
    COLOR_UNTRACKED = untracked;
  }

  let modified: any = config.get("gitstatusbg.modifiedFileBackground");

  if (modified) {
    COLOR_MODIFIED = modified;
  }

  let behind: any = config.get("gitstatusbg.behindRemoteFileBackground");

  if (behind) {
    COLOR_BEHIND = behind;
  }
}

async function SetColorToStatus(
  editor?: vscode.TextEditor,
  range?: vscode.Range[]
) {
  let rangeToUse;
  if (range) {
    rangeToUse = range;
  } else {
    rangeToUse = editor?.visibleRanges;
  }

  if (!rangeToUse || !editor || !isTextEditor(editor)) {
    return;
  }

  let status = await GetStatus(editor.document.uri.fsPath);

  if (status) {
    let bgcolor = undefined;

    switch (status) {
      case Status.MODIFIED: {
        bgcolor = COLOR_MODIFIED;
        break;
      }
      case Status.UNTRACKED: {
        bgcolor = COLOR_UNTRACKED;
        break;
      }
      case XStatus.CHANGED_ON_SERVER: {
        bgcolor = COLOR_BEHIND;
        break;
      }
    }

    if (bgcolor) {
      ColorBackground(bgcolor, editor, rangeToUse);
    }
  }
}

async function GetStatus(_fspath: string) {
  if (!git) {
    return undefined;
  }

  let repo = git.repositories[0];

  if (!repo) {
    return undefined;
  }

  var cached = fileStatusCache[_fspath];
  if (cached) {
    return cached;
  }

  await repo.status();

  let result = undefined;

  repo.state.workingTreeChanges.forEach((value, _index, _changes) => {
    if (value.uri.fsPath === _fspath) {
      switch (value.status) {
        case Status.MODIFIED:
        case Status.UNTRACKED: {
          result = value.status;
          fileStatusCache[_fspath] = result;
          break;
        }
      }
    }
  });

  if (!result) {
    let behind = repo.state.HEAD?.behind;
    if (!behind ? false : behind > 0) {
      // There are changes on the server
      let upstream = repo.state.HEAD?.upstream;
      if (upstream && upstream.name && upstream.remote) {
        let upstreamDiff = await repo.diffWith(
          `${upstream?.remote}/${upstream?.name}`
        );

        upstreamDiff.forEach((value, _index, _changes) => {
          if (value.uri.fsPath === _fspath) {
            result = XStatus.CHANGED_ON_SERVER;
            fileStatusCache[_fspath] = result;
          }
        });
      }
    }
  }

  return result;
}

function ColorBackground(
  color: string,
  editor?: vscode.TextEditor,
  range?: vscode.Range[]
) {
  if (!range || !editor || !isTextEditor(editor)) {
    return;
  }

  var existing = currentDecorations[editor.document.fileName];

  if (existing) {
    editor.setDecorations(existing, []);
  }

  var myDecorator = vscode.window.createTextEditorDecorationType({
    backgroundColor: color,
    isWholeLine: true,
  });

  editor.setDecorations(myDecorator, range);

  currentDecorations[editor.document.fileName] = myDecorator;
}

// Adapted from https://github.com/cschleiden/vscode-github-actions/blob/master/src/tracker/workflowDocumentTracker.ts
enum DocumentSchemes {
  DebugConsole = "debug",
  Output = "output",
}

function isTextEditor(editor: vscode.TextEditor): boolean {
  const scheme = editor.document.uri.scheme;
  return (
    scheme !== DocumentSchemes.Output && scheme !== DocumentSchemes.DebugConsole
  );
}

// this method is called when your extension is deactivated
export function deactivate() {}


out/extension.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
// alpha of 0.05(5%) should allow for subtle overlay on dark or light backgrounds
// These are also defined in the configuration
let COLOR_MODIFIED = "rgba(255, 165, 0, 0.05)";
let COLOR_UNTRACKED = "rgba(71, 255, 25, 0.05)";
let COLOR_BEHIND = "rgba(255, 0, 0, 0.05)";
let git = undefined;
// Keep track of current decorations so that they can be removed
let currentDecorations = {};
let fileStatusCache = {};
function activate(context) {
    return __awaiter(this, void 0, void 0, function* () {
        const ext = vscode.extensions.getExtension("vscode.git");
        yield (ext === null || ext === void 0 ? void 0 : ext.activate());
        git = ext.exports.getAPI(1);
        git.onDidChangeState;
        context.subscriptions.push(
        // Handle switching (or opening) document tabs
        vscode.window.onDidChangeActiveTextEditor(onDidChangeActiveTextEditor), 
        // Handle scrolling in the document.
        vscode.window.onDidChangeTextEditorVisibleRanges(onDidChangeTextEditorVisibleRanges), 
        // File saved so pick up any changes because of saving
        vscode.workspace.onDidSaveTextDocument(onFileSaved), 
        // Potentially large number of files changed - just reset the status cache (lazy but effective)
        vscode.workspace.onDidChangeWorkspaceFolders((e) => {
            fileStatusCache = {};
        }), vscode.workspace.onDidCreateFiles((e) => {
            fileStatusCache = {};
        }), vscode.workspace.onDidDeleteFiles((e) => {
            fileStatusCache = {};
        }), vscode.workspace.onDidRenameFiles((e) => {
            fileStatusCache = {};
        }));
        // Check for initial document
        RefreshCurrentEditor();
    });
}
exports.activate = activate;
function RefreshCurrentEditor() {
    onDidChangeActiveTextEditor(vscode.window.activeTextEditor);
}
function onFileSaved(e) {
    var _a, _b;
    // Document may have been colored before saving but shouldn't be colored after
    // - therefore remove any existing coloring
    var existing = currentDecorations[e.fileName];
    if (existing) {
        (_b = (_a = vscode.window) === null || _a === void 0 ? void 0 : _a.activeTextEditor) === null || _b === void 0 ? void 0 : _b.setDecorations(existing, []);
    }
    // Delete the cached status because it may have changed because saving
    fileStatusCache[e.fileName] = undefined;
    // Update the coloring of the document
    SetColorToStatus(vscode.window.activeTextEditor);
}
function onDidChangeTextEditorVisibleRanges(e) {
    let changedRanges = [];
    changedRanges = changedRanges.concat(e.visibleRanges);
    SetColorToStatus(e.textEditor, changedRanges);
}
function onDidChangeActiveTextEditor(editor) {
    if (editor === undefined) {
        return;
    }
    ReloadConfigurationInCaseChanged();
    // Clear the cached state of the file just switched to
    // - this allows for state changes of open files while another tab was displayed
    fileStatusCache[editor.document.fileName] = undefined;
    SetColorToStatus(editor);
}
function ReloadConfigurationInCaseChanged() {
    var config = vscode.workspace.getConfiguration();
    let untracked = config.get("gitstatusbg.untrackedFileBackground");
    if (untracked) {
        COLOR_UNTRACKED = untracked;
    }
    let modified = config.get("gitstatusbg.modifiedFileBackground");
    if (modified) {
        COLOR_MODIFIED = modified;
    }
    let behind = config.get("gitstatusbg.behindRemoteFileBackground");
    if (behind) {
        COLOR_BEHIND = behind;
    }
}
function SetColorToStatus(editor, range) {
    return __awaiter(this, void 0, void 0, function* () {
        let rangeToUse;
        if (range) {
            rangeToUse = range;
        }
        else {
            rangeToUse = editor === null || editor === void 0 ? void 0 : editor.visibleRanges;
        }
        if (!rangeToUse || !editor || !isTextEditor(editor)) {
            return;
        }
        let status = yield GetStatus(editor.document.uri.fsPath);
        if (status) {
            let bgcolor = undefined;
            switch (status) {
                case 5 /* MODIFIED */: {
                    bgcolor = COLOR_MODIFIED;
                    break;
                }
                case 7 /* UNTRACKED */: {
                    bgcolor = COLOR_UNTRACKED;
                    break;
                }
                case 100 /* CHANGED_ON_SERVER */: {
                    bgcolor = COLOR_BEHIND;
                    break;
                }
            }
            if (bgcolor) {
                ColorBackground(bgcolor, editor, rangeToUse);
            }
        }
    });
}
function GetStatus(_fspath) {
    var _a, _b;
    return __awaiter(this, void 0, void 0, function* () {
        if (!git) {
            return undefined;
        }
        let repo = git.repositories[0];
        if (!repo) {
            return undefined;
        }
        var cached = fileStatusCache[_fspath];
        if (cached) {
            return cached;
        }
        yield repo.status();
        let result = undefined;
        repo.state.workingTreeChanges.forEach((value, _index, _changes) => {
            if (value.uri.fsPath === _fspath) {
                switch (value.status) {
                    case 5 /* MODIFIED */:
                    case 7 /* UNTRACKED */: {
                        result = value.status;
                        fileStatusCache[_fspath] = result;
                        break;
                    }
                }
            }
        });
        if (!result) {
            let behind = (_a = repo.state.HEAD) === null || _a === void 0 ? void 0 : _a.behind;
            if (!behind ? false : behind > 0) {
                // There are changes on the server
                let upstream = (_b = repo.state.HEAD) === null || _b === void 0 ? void 0 : _b.upstream;
                if (upstream && upstream.name && upstream.remote) {
                    let upstreamDiff = yield repo.diffWith(`${upstream === null || upstream === void 0 ? void 0 : upstream.remote}/${upstream === null || upstream === void 0 ? void 0 : upstream.name}`);
                    upstreamDiff.forEach((value, _index, _changes) => {
                        if (value.uri.fsPath === _fspath) {
                            result = 100 /* CHANGED_ON_SERVER */;
                            fileStatusCache[_fspath] = result;
                        }
                    });
                }
            }
        }
        return result;
    });
}
function ColorBackground(color, editor, range) {
    if (!range || !editor || !isTextEditor(editor)) {
        return;
    }
    var existing = currentDecorations[editor.document.fileName];
    if (existing) {
        editor.setDecorations(existing, []);
    }
    var myDecorator = vscode.window.createTextEditorDecorationType({
        backgroundColor: color,
        isWholeLine: true,
    });
    editor.setDecorations(myDecorator, range);
    currentDecorations[editor.document.fileName] = myDecorator;
}
// Adapted from https://github.com/cschleiden/vscode-github-actions/blob/master/src/tracker/workflowDocumentTracker.ts
var DocumentSchemes;
(function (DocumentSchemes) {
    DocumentSchemes["DebugConsole"] = "debug";
    DocumentSchemes["Output"] = "output";
})(DocumentSchemes || (DocumentSchemes = {}));
function isTextEditor(editor) {
    const scheme = editor.document.uri.scheme;
    return (scheme !== DocumentSchemes.Output && scheme !== DocumentSchemes.DebugConsole);
}
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map